tests/python/kivaloo/wire.py

```diff
@@ -0,0 +1,174 @@
+_CRC_FUNC = crcmod.mkCrcFun(_CASTAGNOLI, initCrc=_ILONGATSAC, rev=True)
+
+
+class NotBinaryData(TypeError):
+    """ You must pass some `bytes` to this function. """
+    pass
+
+
+def checksum(data):
+    """ Calculate the reversed CRC32C checksum on the data. """
+    crc = _CRC_FUNC(data)
+    # swap bytes in that 32-bit value (which shouldn't be interpreted as a
+    # single number)
+    out = 0
+    for pos in [0, 8, 16, 24]:
+        hexit = (crc >> pos) & 0xff
+        out += hexit << (24 - pos)
+    return out
+
+
+def make_packet(idnum, record):
+    """ Create a Wire packet from a record. """
+    record_length = len(record)
+
+    # Calculate CRC32c on first 12 bytes
+    crc_header = checksum(struct.pack('>QI', idnum, record_length))
+
+    # Calculate CRC32 on record, xor CRC32 on first 12 bytes
+    crc_record = checksum(record) ^ crc_header
+
+    # packet
+    packet = struct.pack('>QII%dsI' % record_length,
+                         idnum, record_length, crc_header, record, crc_record)
+    return packet
+
+
+def split_packet(packet):
+    """ Split a Wire packet into a record. """
+    idnum, record_length, crc_recv = struct.unpack_from('>QII', packet)
+
+    # Check CRC32 on first 12 bytes
+    crc_first = checksum(struct.pack('>QI', idnum, record_length))
+    if crc_recv != crc_first:
+        raise Exception("Mismatch in header checksums")
+
+    # Get main message.  16 = 8+4+4 = len(QII).
+    record, crc = struct.unpack_from('>%dsI' % record_length,
+                                     packet, offset=16)
+
+    crc_record = checksum(struct.pack('>%ds' % record_length, record))
+    if crc != (crc_record ^ crc_first):
+        raise Exception("Mismatch in record checksums")
+
+    # Sanity check
+    assert len(packet) == len(record) + 16 + 4
+
+    return idnum, record
+
+
+class Wire(object):
+    """ Base class for the kivaloo Wire protocol. """
+    def __init__(self):
+        self.msgnum = 0
+        self.socket_name = None
+        self.sock = None
+
+    def __del__(self):
+        self.close()
+
+    def _start(self):
+        """ Connect to a server. """
+        self.sock = socket.socket(socket.AF_UNIX)
+        self.sock.connect(self.socket_name)
+
+    def close(self):
+        """ Close a connection to a server. """
+        if self.sock:
+            self.sock.close()
+        self.sock = None
+
+    def send(self, record):
+        """ Send a message. """
+        data = make_packet(self.msgnum, record)
+        self.msgnum += 1
+        self.sock.sendall(data)
+
+    def send_recv_block(self, record):
+        """ Send a message and receive the response.  Supports responses up
+            to wire.MAX_RESPONSE_BYTES. """
+        data = make_packet(self.msgnum, record)
+        self.msgnum += 1
+        self.sock.sendall(data)
+
+        idnum, record = split_packet(self.sock.recv(MAX_RESPONSE_BYTES))
+        assert idnum == self.msgnum - 1
+
+        return record
+
+
+class Wire_lbs(Wire):
```

Comment:
Is there a reason you didn't call this `proto_lbs`?  In the kivaloo code I distinguish between the `wire` level (opaque packets) and the `proto` level (packets have meanings as requests and responses).